File: src/c_two/_native/foundation/c2-config/src/ipc.rs
//! IPC transport configuration split into Base / Server / Client.
//!
//! `BaseIpcConfig` holds pool, chunk, and reassembly settings shared by both
//! sides.  `ServerIpcConfig` and `ClientIpcConfig` each wrap a `base` field
//! and add role-specific knobs.  Both implement `Deref<Target = BaseIpcConfig>`
//! for ergonomic field access.
//!
//! All size fields are in bytes; time fields are in seconds (with `_secs`
//! suffix).  `Default` impls exist for Rust-side testing only — Python is
//! authoritative at runtime.

use std::ops::Deref;

// ─── Base ────────────────────────────────────────────────────────────────────

/// Fields shared by both server and client IPC configs.
#[derive(Debug, Clone)]
pub struct BaseIpcConfig {
    // ── Pool SHM settings ────────────────────────────────────────────────
    pub pool_enabled: bool,
    pub pool_segment_size: u64,
    pub max_pool_segments: u32,
    pub max_pool_memory: u64,

    // ── Reassembly pool settings ─────────────────────────────────────────
    pub reassembly_segment_size: u64,
    pub reassembly_max_segments: u32,

    // ── Chunked transfer settings ────────────────────────────────────────
    pub max_total_chunks: u32,
    pub chunk_gc_interval_secs: f64,
    pub chunk_threshold_ratio: f64,
    pub chunk_assembler_timeout_secs: f64,
    pub max_reassembly_bytes: u64,
    pub chunk_size: u64,
}

// ─── Server ──────────────────────────────────────────────────────────────────

/// Server-side IPC configuration.
#[derive(Debug, Clone)]
pub struct ServerIpcConfig {
    pub base: BaseIpcConfig,

    pub shm_threshold: u64,
    pub max_frame_size: u64,
    pub max_payload_size: u64,
    pub max_pending_requests: u32,
    pub pool_decay_seconds: f64,
    pub heartbeat_interval_secs: f64,
    pub heartbeat_timeout_secs: f64,
}

// ─── Client ──────────────────────────────────────────────────────────────────

/// Client-side IPC configuration.
#[derive(Debug, Clone)]
pub struct ClientIpcConfig {
    pub base: BaseIpcConfig,
    pub shm_threshold: u64,
}

// ─── Default impls ───────────────────────────────────────────────────────────

impl Default for BaseIpcConfig {
    fn default() -> Self {
        Self {
            pool_enabled: true,
            pool_segment_size: 268_435_456,       // 256 MB
            max_pool_segments: 4,
            max_pool_memory: 1_073_741_824,        // 1 GB

            reassembly_segment_size: 268_435_456,  // 256 MB (server default)
            reassembly_max_segments: 4,

            max_total_chunks: 512,
            chunk_gc_interval_secs: 5.0,
            chunk_threshold_ratio: 0.9,
            chunk_assembler_timeout_secs: 60.0,
            max_reassembly_bytes: 8_589_934_592,   // 8 GB
            chunk_size: 131_072,                   // 128 KB
        }
    }
}

impl Default for ServerIpcConfig {
    fn default() -> Self {
        Self {
            base: BaseIpcConfig::default(),

            shm_threshold: 4_096,
            max_frame_size: 2_147_483_648,         // 2 GB
            max_payload_size: 17_179_869_184,      // 16 GB
            max_pending_requests: 1024,
            pool_decay_seconds: 60.0,
            heartbeat_interval_secs: 15.0,
            heartbeat_timeout_secs: 30.0,
        }
    }
}

impl Default for ClientIpcConfig {
    fn default() -> Self {
        Self {
            base: BaseIpcConfig {
                reassembly_segment_size: 64 * 1024 * 1024, // 64 MB override
                ..BaseIpcConfig::default()
            },
            shm_threshold: 4_096,
        }
    }
}

// ─── Deref impls ─────────────────────────────────────────────────────────────

impl Deref for ServerIpcConfig {
    type Target = BaseIpcConfig;
    fn deref(&self) -> &BaseIpcConfig { &self.base }
}

impl Deref for ClientIpcConfig {
    type Target = BaseIpcConfig;
    fn deref(&self) -> &BaseIpcConfig { &self.base }
}

// ─── Validation ──────────────────────────────────────────────────────────────

impl BaseIpcConfig {
    pub fn validate(&self) -> Result<(), String> {
        if self.pool_segment_size == 0 {
            return Err("pool_segment_size must be > 0".into());
        }
        if self.pool_segment_size > u32::MAX as u64 {
            return Err(format!(
                "pool_segment_size ({}) must be <= u32::MAX ({}) for wire format compatibility",
                self.pool_segment_size, u32::MAX,
            ));
        }
        if self.chunk_size == 0 {
            return Err("chunk_size must be > 0".into());
        }
        if self.chunk_threshold_ratio <= 0.0 || self.chunk_threshold_ratio > 1.0 {
            return Err(format!(
                "chunk_threshold_ratio ({}) must be in (0, 1]",
                self.chunk_threshold_ratio,
            ));
        }
        Ok(())
    }
}

impl ServerIpcConfig {
    pub fn validate(&self) -> Result<(), String> {
        self.base.validate()?;
        if self.max_frame_size <= 16 {
            return Err(format!(
                "max_frame_size ({}) must be > 16 (header size)",
                self.max_frame_size,
            ));
        }
        if self.max_payload_size == 0 {
            return Err("max_payload_size must be > 0".into());
        }
        if self.shm_threshold > self.max_frame_size {
            return Err(format!(
                "shm_threshold ({}) must not exceed max_frame_size ({})",
                self.shm_threshold, self.max_frame_size,
            ));
        }
        Ok(())
    }
}

impl ClientIpcConfig {
    pub fn validate(&self) -> Result<(), String> {
        self.base.validate()
    }
}

// ─── Tests ───────────────────────────────────────────────────────────────────

#[cfg(test)]
mod tests {
    use super::*;

    // ── Default validation ───────────────────────────────────────────────

    #[test]
    fn server_default_validates() {
        assert!(ServerIpcConfig::default().validate().is_ok());
    }

    #[test]
    fn client_default_validates() {
        assert!(ClientIpcConfig::default().validate().is_ok());
    }

    #[test]
    fn base_default_validates() {
        assert!(BaseIpcConfig::default().validate().is_ok());
    }

    // ── Base validations (via ServerIpcConfig) ───────────────────────────

    #[test]
    fn reject_zero_segment_size() {
        let mut cfg = ServerIpcConfig::default();
        cfg.base.pool_segment_size = 0;
        assert!(cfg.validate().unwrap_err().contains("pool_segment_size must be > 0"));
    }

    #[test]
    fn reject_oversized_segment() {
        let mut cfg = ServerIpcConfig::default();
        cfg.base.pool_segment_size = u64::from(u32::MAX) + 1;
        assert!(cfg.validate().unwrap_err().contains("u32::MAX"));
    }

    #[test]
    fn reject_zero_chunk_size() {
        let mut cfg = ServerIpcConfig::default();
        cfg.base.chunk_size = 0;
        assert!(cfg.validate().unwrap_err().contains("chunk_size must be > 0"));
    }

    #[test]
    fn reject_bad_threshold_ratio() {
        let mut cfg = BaseIpcConfig::default();
        cfg.chunk_threshold_ratio = 0.0;
        assert!(cfg.validate().unwrap_err().contains("chunk_threshold_ratio"));

        cfg.chunk_threshold_ratio = 1.1;
        assert!(cfg.validate().unwrap_err().contains("chunk_threshold_ratio"));

        cfg.chunk_threshold_ratio = 1.0; // boundary — valid
        assert!(cfg.validate().is_ok());
    }

    // ── Server-only validations ──────────────────────────────────────────

    #[test]
    fn reject_small_frame_size() {
        let mut cfg = ServerIpcConfig::default();
        cfg.max_frame_size = 16;
        assert!(cfg.validate().unwrap_err().contains("max_frame_size"));
    }

    #[test]
    fn reject_zero_payload_size() {
        let mut cfg = ServerIpcConfig::default();
        cfg.max_payload_size = 0;
        assert!(cfg.validate().unwrap_err().contains("max_payload_size must be > 0"));
    }

    #[test]
    fn reject_threshold_exceeds_frame() {
        let mut cfg = ServerIpcConfig::default();
        cfg.shm_threshold = 1000;
        cfg.max_frame_size = 500;
        assert!(cfg.validate().unwrap_err().contains("shm_threshold"));
    }

    // ── Deref ────────────────────────────────────────────────────────────

    #[test]
    fn server_deref_accesses_base_fields() {
        let cfg = ServerIpcConfig::default();
        assert!(cfg.pool_enabled);
        assert_eq!(cfg.pool_segment_size, 268_435_456);
    }

    #[test]
    fn client_deref_accesses_base_fields() {
        let cfg = ClientIpcConfig::default();
        assert!(cfg.pool_enabled);
        assert_eq!(cfg.chunk_size, 131_072);
    }

    // ── Client-specific defaults ─────────────────────────────────────────

    #[test]
    fn client_reassembly_segment_is_64mb() {
        let cfg = ClientIpcConfig::default();
        assert_eq!(cfg.base.reassembly_segment_size, 64 * 1024 * 1024);
    }

    #[test]
    fn server_reassembly_segment_is_256mb() {
        let cfg = ServerIpcConfig::default();
        assert_eq!(cfg.base.reassembly_segment_size, 268_435_456);
    }

    // ── Type checks ──────────────────────────────────────────────────────

    #[test]
    fn chunk_gc_interval_secs_is_f64() {
        let cfg = ServerIpcConfig::default();
        let _secs: f64 = cfg.chunk_gc_interval_secs;
        assert!((cfg.chunk_gc_interval_secs - 5.0).abs() < f64::EPSILON);
    }

    #[test]
    fn chunk_size_is_u64() {
        let cfg = BaseIpcConfig::default();
        let _v: u64 = cfg.chunk_size;
        assert_eq!(_v, 131_072);
    }

    // ── Client validate delegates to base ────────────────────────────────

    #[test]
    fn client_rejects_zero_segment_size() {
        let mut cfg = ClientIpcConfig::default();
        cfg.base.pool_segment_size = 0;
        assert!(cfg.validate().unwrap_err().contains("pool_segment_size must be > 0"));
    }
}

File: sdk/python/src/c_two/config/ipc.py
"""Frozen IPC configuration dataclasses with validation."""
from __future__ import annotations

from dataclasses import dataclass, fields
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from .settings import C2Settings


@dataclass(frozen=True)
class BaseIPCConfig:
    """Shared IPC configuration for both server and client."""

    pool_enabled: bool = True
    pool_segment_size: int = 268_435_456        # 256 MB
    max_pool_segments: int = 4
    max_pool_memory: int = 1_073_741_824        # 1 GB
    reassembly_segment_size: int = 268_435_456  # 256 MB
    reassembly_max_segments: int = 4
    max_total_chunks: int = 512
    chunk_gc_interval: float = 5.0
    chunk_threshold_ratio: float = 0.9
    chunk_assembler_timeout: float = 60.0
    max_reassembly_bytes: int = 8_589_934_592   # 8 GB
    chunk_size: int = 131_072                   # 128 KB

    def __post_init__(self) -> None:
        if self.pool_segment_size <= 0 or self.pool_segment_size > 0xFFFFFFFF:
            raise ValueError(
                f'pool_segment_size must be in (0, {0xFFFFFFFF}], '
                f'got {self.pool_segment_size}'
            )
        if not 1 <= self.max_pool_segments <= 255:
            raise ValueError(
                f'max_pool_segments must be 1..255, got {self.max_pool_segments}'
            )
        if not (0 < self.chunk_threshold_ratio <= 1):
            raise ValueError(
                f'chunk_threshold_ratio must be in (0, 1], '
                f'got {self.chunk_threshold_ratio}'
            )
        if self.chunk_gc_interval <= 0:
            raise ValueError(
                f'chunk_gc_interval must be positive, got {self.chunk_gc_interval}'
            )
        if self.chunk_assembler_timeout <= 0:
            raise ValueError(
                f'chunk_assembler_timeout must be positive, '
                f'got {self.chunk_assembler_timeout}'
            )
        if not 1 <= self.reassembly_max_segments <= 255:
            raise ValueError(
                f'reassembly_max_segments must be 1..255, '
                f'got {self.reassembly_max_segments}'
            )
        if self.reassembly_segment_size <= 0:
            raise ValueError(
                f'reassembly_segment_size must be positive, '
                f'got {self.reassembly_segment_size}'
            )


@dataclass(frozen=True)
class ServerIPCConfig(BaseIPCConfig):
    """Server-side IPC configuration."""

    max_frame_size: int = 2_147_483_648         # 2 GB
    max_payload_size: int = 17_179_869_184      # 16 GB
    max_pending_requests: int = 1024
    pool_decay_seconds: float = 60.0
    heartbeat_interval: float = 15.0
    heartbeat_timeout: float = 30.0

    def __post_init__(self) -> None:
        super().__post_init__()
        if self.max_frame_size <= 16:
            raise ValueError(
                f'max_frame_size must be > 16, got {self.max_frame_size}'
            )
        if self.max_payload_size <= 0:
            raise ValueError(
                f'max_payload_size must be positive, got {self.max_payload_size}'
            )
        if self.heartbeat_interval < 0:
            raise ValueError(
                f'heartbeat_interval must be >= 0, got {self.heartbeat_interval}'
            )
        if self.heartbeat_interval > 0 and self.heartbeat_timeout <= self.heartbeat_interval:
            raise ValueError(
                f'heartbeat_timeout ({self.heartbeat_timeout}) must exceed '
                f'heartbeat_interval ({self.heartbeat_interval})'
            )
        if self.max_pool_memory < self.pool_segment_size:
            raise ValueError(
                f'max_pool_memory ({self.max_pool_memory}) must be >= '
                f'pool_segment_size ({self.pool_segment_size})'
            )


@dataclass(frozen=True)
class ClientIPCConfig(BaseIPCConfig):
    """Client-side IPC configuration."""

    reassembly_segment_size: int = 67_108_864   # 64 MB


def build_server_config(
    settings: C2Settings | None = None,
    **kwargs: object,
) -> ServerIPCConfig:
    """Build a ``ServerIPCConfig`` from kwargs > env vars > class defaults."""
    if settings is None:
        from .settings import settings as _settings
        settings = _settings

    merged: dict[str, object] = {}
    for f in fields(ServerIPCConfig):
        name = f.name
        env_name = f'ipc_{name}'
        if name in kwargs:
            merged[name] = kwargs[name]
        elif hasattr(settings, env_name) and getattr(settings, env_name) is not None:
            merged[name] = getattr(settings, env_name)

    # Clamping: pool_segment_size <= max_payload_size
    seg = merged.get('pool_segment_size', ServerIPCConfig.pool_segment_size)
    pay = merged.get('max_payload_size', ServerIPCConfig.max_payload_size)
    if seg > pay:  # type: ignore[operator]
        merged['pool_segment_size'] = pay

    # Auto-derive max_pool_memory if not explicitly set
    if 'max_pool_memory' not in merged:
        seg_final = merged.get('pool_segment_size', ServerIPCConfig.pool_segment_size)
        segs = merged.get('max_pool_segments', ServerIPCConfig.max_pool_segments)
        merged['max_pool_memory'] = seg_final * segs  # type: ignore[operator]

    return ServerIPCConfig(**merged)


def build_client_config(
    settings: C2Settings | None = None,
    **kwargs: object,
) -> ClientIPCConfig:
    """Build a ``ClientIPCConfig`` from kwargs > env vars > class defaults."""
    if settings is None:
        from .settings import settings as _settings
        settings = _settings

    merged: dict[str, object] = {}
    for f in fields(ClientIPCConfig):
        name = f.name
        env_name = f'ipc_{name}'
        if name in kwargs:
            merged[name] = kwargs[name]
        elif hasattr(settings, env_name) and getattr(settings, env_name) is not None:
            merged[name] = getattr(settings, env_name)

    return ClientIPCConfig(**merged)
